### DISCUSSION TITLE
Error when using PADL2Loss

### DISCUSSION THREAD

**Discussion by Ubivision** [yoshitomo-matsubara/torchdistill#239] [Q&A] [answered]
I run wide_resnet40_1_from_wide_resnet40_4-hyperparameter_tuning.yaml directly and there is no problem, but when I modify the type of criterion in this config to PADL2Loss, an error occurs. The following is the error message:
`TypeError
__init__() missing 2 required positional arguments: 'student_embed_module_path' and 'teacher_embed_module_path'`
This error occurs at line 111 get_distillation_box() in image_classification.py
After checking the error, I found that the init parameters of the KDLoss class and the PADL2Loss class in single.py are different, which caused this problem. Is this a problem with my config or something else?

**Comment by yoshitomo-matsubara** [comment, marked as answer]
Hi @Ubivision 

As the error message says, your yaml config misses required parameters to instantiate `PADL2Loss`.
I assume you're using a config file for standard KD, so I would suggest you compare configs for KD (single-stage training) and PAD+L2 (two-stage training).
E.g., Loss classes [`KDLoss`](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/ilsvrc2012/single_stage/kd/resnet18_from_resnet152.yaml#L94-L104) vs. [`PADL2Loss`](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/ilsvrc2012/multi_stage/pad/resnet18_from_resnet34.yaml#L150-L171)

Note that PAD + L2 uses multiple intermediate outputs from student and teachers e.g., `student_embed_module_path: 'student_model.avgpool'` (you could imagine `resnet18` as `student_model` in this case, where `student_model.avgpool` means `resnet18.avgpool`), and these module paths vary from model to model.

If you want to use CIFAR-10/100 for quick experiments, I would suggest you start with [this yaml file](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/ilsvrc2012/multi_stage/pad/resnet18_from_resnet34.yaml) for ImageNet dataset and replace dataset, teacher, and student entries with those you want to use e.g., cifar10, wide_resnet40_4, and wide_resnet40_1.
Also as mentioned [here](https://github.com/yoshitomo-matsubara/torchdistill/tree/main/configs#sample), most of the configs under `sample/` are provided for debugging purpose or as starter configs to design your experiments. So you will need to tune hyperparameters in the configs by yourself.

**Reply by Ubivision** [reply]
Thanks! I ran the PAD successfully as you suggested
